### PR TITLE
[cni-cilium] Full distroless

### PR DESCRIPTION
**Pull request by AndreyPavlovFlant**
## Description
Cilium distroless images.

## Why do we need it, and what problem does it solve?
Images with only necessary bins are safer.

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: cni-cilium
type: chore
summary: Distroless images.
impact: All cilium-related pods will restart. L7 and FQDN-based policies will flap.
impact_level: high
```

### CODE REVIEW

File: modules/021-cni-cilium/images/bin-cilium/werf.inc.yaml [hunk 1]
```diff
@@ -47,15 +43,13 @@ shell:
...
-  - dpkg -i /iptables/*.deb
-  - rm -rf /iptables
   - chmod +x /go/src/github.com/cilium/cilium/images/runtime/*.sh
   - cd /go/src/github.com/cilium/cilium/images/runtime
   - mkdir -p orig && cp -a iptables-wrapper-installer.sh orig/
   - ./iptables-wrapper-installer.sh --no-sanity-check
   setup:
   - export GOROOT=/usr/local/go GOPATH=/go
-  - export PATH=${GOROOT}/bin:${GOPATH}/bin:${PATH}
+  - export PATH=${GOROOT}/bin:${GOPATH}/bin:/root/bin:/sbin:/usr/sbin:/usr/local/sbin:/bin:/usr/bin:/usr/local/bin
```
Comment (apolovov) [inline modules/021-cni-cilium/images/bin-cilium/werf.inc.yaml:52]: A comment, why we configure PATH explicitly.

File: modules/021-cni-cilium/images/bin-iptables/werf.inc.yaml [hunk 2]
```diff
@@ -8,15 +8,30 @@
...
 # and https://salsa.debian.org/pkg-netfilter-team/pkg-iptables/-/tree/debian/1.8.8-1
+# and https://git.netfilter.org/iptables/tree/?h=v1.8.8
+# and https://packages.altlinux.org/ru/p10/srpms/iptables/specfiles/2623143627906179825
 # #####################################################################
 ---
 artifact: {{ $.ModuleName }}/iptables-artifact
 fromImage: {{ $.ModuleName }}/base-cilium-dev
 shell:
+  beforeInstall:
+    - apt-get remove --purge -y iptables iptables-ipv6 iptables-nft
```
Comment (apolovov) [inline modules/021-cni-cilium/images/bin-iptables/werf.inc.yaml:19]:
```suggestion
    - apt-get remove --purge -y iptables iptables-ipv6 iptables-nft # JIC
```

File: modules/021-cni-cilium/images/base-cilium-dev/werf.inc.yaml [hunk 3]
```diff
@@ -25,7 +25,7 @@
 # #####################################################################
```
Comment (apolovov) [inline modules/021-cni-cilium/images/base-cilium-dev/werf.inc.yaml:25]: Need some TODO about moving the image to basic d8 images.

File: modules/021-cni-cilium/images/agent/werf.inc.yaml [hunk 4]
```diff
@@ -8,45 +8,57 @@
...
 # for prepull
-{{ $binaries := cat $binaries "/pause /usr/bin/true" }}
----
+{{ $binariesFromALT := cat $binariesFromALT "/bin/true" }}
+# for debug
+{{ $binariesFromALT := cat $binariesFromALT "/bin/sleep /bin/ls /usr/bin/curl /usr/bin/ldd /usr/bin/strace" }}
+#####################################################################
+{{ $selfBuildedBinaries := "" }}
+# from llvm-artifact
+{{ $selfBuildedBinaries := cat $selfBuildedBinaries "/usr/local/bin/clang /usr/local/bin/llc" }}
```
Comment (apolovov) [inline modules/021-cni-cilium/images/agent/werf.inc.yaml:39]: Builded -> Built 📖.

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/openstack'...

**Comment by AndreyPavlovFlant** [comment]
/e2e/abort openstack;Standard;containerd;1.27 9716567592 9716567592-2-con-1-27 /sys/deckhouse-oss/install:pr8636 redos@31.129.32.211

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by AndreyPavlovFlant** [comment]
/e2e/abort static;Static;containerd;1.27 9716567038 9716567038-1-con-1-27 /sys/deckhouse-oss/install:pr8636 astra@192.168.199.50

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/static'...

**Comment by z9r5** [comment]
/e2e/abort openstack;Standard;containerd;1.27 9716567592 9716567592-1-con-1-27 /sys/deckhouse-oss/install:pr8636 redos@45.12.231.155

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by z9r5** [comment]
/e2e/abort static;Static;containerd;1.27 9716567038 9716567038-1-con-1-27 /sys/deckhouse-oss/install:pr8636 astra@192.168.199.50

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've recognized your '/e2e/abort' command and started the workflow...

**Review by apolovov — APPROVED** [review]

File: modules/021-cni-cilium/images/agent/werf.inc.yaml [hunk 5]
```diff
@@ -8,45 +8,56 @@
...
-{{ $binaries := cat $binaries "/usr/local/bin/clang /usr/local/bin/llc" }}
-{{ $binaries := cat $binaries "/usr/lib/x86_64-linux-gnu/libstdc++.so.6" }}
-# bpftool and dependencies
-{{ $binaries := cat $binaries "/usr/local/bin/bpftool" }}
-# cni-loopback and dependencies
-{{ $binaries := cat $binaries "/cni/loopback" }}
-# gops and dependencies
-{{ $binaries := cat $binaries "/bin/gops" }}
+## ln -snf /usr/share/zoneinfo/Etc/UTC /etc/localtime
+{{ $binariesFromALT := "/etc/localtime /usr/bin/jq" }}
```
Comment (RomanenkoDenys) [inline modules/021-cni-cilium/images/agent/werf.inc.yaml:12]: we have statically builded JQ binary in registry packages

File: modules/021-cni-cilium/images/agent/werf.inc.yaml [hunk 6]
```diff
@@ -8,45 +8,56 @@
...
-{{ $binaries := cat $binaries "/bin/kmod /bin/lsmod /sbin/depmod /sbin/insmod /sbin/lsmod /sbin/modinfo /sbin/modprobe /sbin/rmmod" }}
+{{ $binariesFromALT := cat $binariesFromALT "/bin/kmod /bin/lsmod /sbin/depmod /sbin/insmod /sbin/lsmod /sbin/modinfo /sbin/modprobe /sbin/rmmod" }}
 # iproute2 and dependencies
-{{ $binaries := cat $binaries "/bin/ip /bin/ss /sbin/ip /sbin/bridge /sbin/dcb /sbin/devlink /sbin/rtacct /sbin/rtmon /sbin/tc /sbin/tipc /sbin/vdpa /usr/bin/lnstat" }}
-{{ $binaries := cat $binaries "/usr/bin/nstat /usr/bin/rdma /usr/bin/routef /usr/bin/routel /usr/bin/ctstat /usr/bin/rtstat /usr/sbin/arpd /usr/sbin/genl" }}
-# iptables and dependencies
-{{ $binaries := cat $binaries "/usr/sbin/xtables* /usr/sbin/arptables* /usr/sbin/ebtables* /usr/sbin/ip6tables* /usr/sbin/iptables* /usr/sbin/ipset* /usr/bin/iptables-xml" }}
-{{ $binaries := cat $binaries "/usr/sbin/nfnl_osf" }}
-{{ $binaries := cat $binaries "/usr/lib/x86_64-linux-gnu/xtables/*" }}
+{{ $binariesFromALT := cat $binariesFromALT "/sbin/ip /usr/sbin/ss /sbin/ip /usr/sbin/bridge /sbin/dcb /sbin/devlink /usr/sbin/rtacct /sbin/rtmon /sbin/tc /usr/sbin/tipc /sbin/vdpa /usr/bin/lnstat" }}
```
Comment (RomanenkoDenys) [inline modules/021-cni-cilium/images/agent/werf.inc.yaml:19]: /sbin/ip two times

File: modules/021-cni-cilium/images/agent/werf.inc.yaml [hunk 7]
```diff
@@ -8,45 +8,56 @@
...
+{{ $binariesFromALT := cat $binariesFromALT "/usr/bin/nstat /sbin/rdma /usr/sbin/routef /usr/sbin/routel /usr/sbin/ctstat /usr/sbin/rtstat /usr/sbin/genl" }}
+{{ $binariesFromALT := cat $binariesFromALT "/usr/sbin/arpd" }}
```
[the remark this answers is not in the document]
Comment (AndreyPavlovFlant) [reply modules/021-cni-cilium/images/agent/werf.inc.yaml]: In cilium, not all basic interactions with the network subsystem are written in golang+ebpf and not in all components. 
Sometimes external programs (like ip, iptables, etc.) are called directly from the go code.

There is no official "distroless" image for Cilium, and there is no list of necessary files to ensure correct operation.  
Therefore, I attempted to transfer all the contents of the packages that were explicitly installed on the original "runtime" image to the "distroless" image. 

The iproute2 package (from ubuntu-jammy) was explicitly installed in the original cilium-agent image.
This package contains the arpd binary file.

Due to the high time costs, each binary file and its aliases were not checked for their use in the code.

I have now re-checked the cilium code and did not find any explicit "arpd" calls. After that, I removed the arpd binary from the distroless image and ran basic tests, which passed without errors.

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/static'...

**Comment by z9r5** [comment]
/e2e/abort static;Static;containerd;1.27 9937782850 9937782850-1-con-1-27 /sys/deckhouse-oss/install:pr8636 astra@192.168.199.220

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/openstack'...

**Comment by AndreyPavlovFlant** [comment]
/e2e/abort openstack;Standard;containerd;1.27 10111728144 10111728144-1-con-1-27 /sys/deckhouse-oss/install:pr8636 redos@45.12.231.206

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/openstack'...

**Comment by AndreyPavlovFlant** [comment]
/e2e/abort openstack;Standard;containerd;1.27 10112524735 10112524735-1-con-1-27 /sys/deckhouse-oss/install:pr8636

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by AndreyPavlovFlant** [comment]
/e2e/abort openstack;Standard;containerd;1.27 10112524735 10112524735-1-con-1-27 /sys/deckhouse-oss/install:pr8636

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/static'...

File: modules/021-cni-cilium/images/base-cilium-dev/werf.inc.yaml.txt [hunk 8]
```diff
@@ -25,44 +25,40 @@
 # #####################################################################
 ---
 image: {{ $.ModuleName }}/base-cilium-dev
-from: {{ $.Images.BASE_UBUNTU }}
-fromCacheVersion: {{ $fromCacheVersion }}
+from: {{ $.Images.BASE_ALT }}
```
Comment (RomanenkoDenys) [inline modules/021-cni-cilium/images/base-cilium-dev/werf.inc.yaml.txt:28]: Why you do not use BASE_ALT_DEV_CILIUM image ?

**Review by RomanenkoDenys — APPROVED** [review]

**Pull request merged**